File: content/snippets/cpp/interview_lcci_01_04_palindrome_permutation.cpp
#include <array>
#include <cctype>
#include <iostream>
#include <string>

bool canPermutePalindrome(const std::string& s) {
    std::array<int, 256> freq{};
    for (unsigned char ch : s) {
        if (std::isspace(ch)) {
            continue;
        }
        unsigned char lower = static_cast<unsigned char>(std::tolower(ch));
        ++freq[lower];
    }

    int odd = 0;
    for (int count : freq) {
        if ((count & 1) == 1) {
            ++odd;
            if (odd > 1) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    std::cout << "tact coa -> " << std::boolalpha << canPermutePalindrome("tact coa") << '\n';
    std::cout << "daily -> " << std::boolalpha << canPermutePalindrome("daily") << '\n';
    return 0;
}

File: content/snippets/typescript/interview_lcci_01_04_palindrome_permutation.ts
function canPermutePalindrome(s: string): boolean {
  const freq = new Map<string, number>();
  for (const ch of s.toLowerCase()) {
    if (ch === " ") {
      continue;
    }
    freq.set(ch, (freq.get(ch) ?? 0) + 1);
  }

  let odd = 0;
  for (const count of freq.values()) {
    if (count % 2 === 1) {
      odd += 1;
      if (odd > 1) {
        return false;
      }
    }
  }
  return true;
}

function main(): void {
  console.log(`tact coa -> ${canPermutePalindrome("tact coa")}`);
  console.log(`daily -> ${canPermutePalindrome("daily")}`);
}

main();
